blue_ai/envs/custom_decay.py

```diff
@@ -0,0 +1,26 @@
+import torch.nn as nn
+import torch
+from torch.nn.modules import MSELoss, Sequential
+
+
+class PositivePenaltyLoss(nn.Module):
+    known = MSELoss()
+
+    """
+    Uses a combinations of MSE Loss with a RELU penalty, meaning that positive 
+    weights will always have a large penalty 
+    """
+
+    def __init__(self, alpha=0.2):
+        super(PositivePenaltyLoss, self).__init__()
+        self.policy_hook: Sequential = None
+        self.alpha = alpha
+
+    def forward(self, inputs: torch.Tensor, targets: torch.Tensor):
+        net = next(self.policy_hook.parameters())
```

Comment:
I think this will only get the first layer. You'll need to iterate through .parameters() to get all the parameters. The trick through, is that we only want to decay the _weights_, not the _biases_. So you should operate on all the 2-D arrays you encounter when iterating through .parameters() but skip over the 1-D arrays... there may be a more efficient way, but that's all I can think of right now.

Reply:
This is as close as I've gotten to getting all the middle layers; I hate it with a passion. But it _works_

```python3
      shape = None
      net = torch.cat(
          [
              x
              for x in self.policy_hook.parameters()
              if x.dim() == 2 and x.shape == (shape or (shape := x.shape))
          ]
      )
```